### PR TITLE
Integrate swap_tensors into nn.Module.load_state_dict

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Added a `torch.Tensor` method that defines how to transform `other`, a value in the state dictionary, to be loaded into `self`, a param/buffer in an `nn.Module` before swapping via `torch.utils.swap_tensors`
* `param.module_load(sd[key])`

This method can be overridden using `__torch_function__`.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117913

### CODE REVIEW

File: test/nn/test_load_state_dict.py [hunk 1]
```diff
@@ -189,7 +191,8 @@ def my_state_dict_pre_hook(*args, **kwargs):
         _ = m.state_dict()
         self.assertTrue(called)
 
-    # FIXME: doesn't fail locally, maybe remove
+    # fails in crossref mode when swapping tensors as LSTM installs weakrefs
+    @skipIfCrossRef
```
Comment (albanD) [inline test/nn/test_load_state_dict.py:195]: Is that independant?

File: test/nn/test_load_state_dict.py [hunk 2]
```diff
@@ -362,7 +365,162 @@ def test_load_state_dict_warn_assign(self):
             m.load_state_dict(state_dict)
 
 
+class MyLoadTensor(torch.Tensor):
+
+    # enable deepcopy
```
Comment (albanD) [inline test/nn/test_load_state_dict.py:370]: deepcopy doesn't work without that?

File: test/nn/test_load_state_dict.py [hunk 3]
```diff
@@ -362,7 +365,162 @@ def test_load_state_dict_warn_assign(self):
...
+        kwargs = {} if kwargs is None else kwargs
+
+        def module_load_from(dest, src):
+            # always convert src to MyLoadTensor
+            if type(src) is torch.Tensor:
+                return MyLoadTensor(src)
+            elif type(src) is MyLoadTensor:
+                return src
+            elif type(src) is MyWrapperLoadTensor:
+                return MyLoadTensor(src._data)
```
Comment (albanD) [inline test/nn/test_load_state_dict.py:385]: What is `._data` ??

File: test/nn/test_load_state_dict.py [hunk 4]
```diff
@@ -362,7 +365,162 @@ def test_load_state_dict_warn_assign(self):
...
+                return src
+            else:
+                raise NotImplementedError(f"module_load_to(): unsupported type(dest) {type(dest)}")
+
+        if func is torch.Tensor.module_load_from:
+            return module_load_from(*args, **kwargs)
+        elif func is torch.Tensor.module_load_to:
+            return module_load_to(*args, **kwargs)
+        elif func is torch.Tensor.detach:
+            return MyLoadTensor(args[0].data.detach())
```
Comment (albanD) [inline test/nn/test_load_state_dict.py:400]: Don't use .data!
Also why is this needed?

File: test/nn/test_load_state_dict.py [hunk 5]
```diff
@@ -362,7 +365,162 @@ def test_load_state_dict_warn_assign(self):
...
+        return t
+
+    def __init__(self, data: torch.Tensor):
+        self._data = data
+
+    def __repr__(self):
+        return f"MyWrapperLoadTensor({self._data.__repr__()})"
+
+    # enable deepcopy
+    def new_empty(self, shape):
```
Comment (albanD) [inline test/nn/test_load_state_dict.py:424]: Similar: why?

File: test/nn/test_load_state_dict.py [hunk 6]
```diff
@@ -362,7 +365,162 @@ def test_load_state_dict_warn_assign(self):
...
+        elif func is torch.Tensor.module_load_to:
+            return module_load_to(*args, **kwargs)
+        elif func is torch.Tensor.detach:
+            return MyLoadTensor(args[0].data.detach())
+        else:
+            with torch._C.DisableTorchFunctionSubclass():
+                return func(*args, **kwargs)
+
+
+class MyWrapperLoadTensor(torch.Tensor):
```
Comment (albanD) [inline test/nn/test_load_state_dict.py:406]: Inherit from MyLoadTensor to reduce duplication?

File: test/nn/test_load_state_dict.py [hunk 7]
```diff
@@ -362,7 +365,162 @@ def test_load_state_dict_warn_assign(self):
...
+            m.register_buffer('buf', torch.randn(2, 3))
+            if subclass is not None:
+                m.weight = torch.nn.Parameter(subclass(m.weight))
+                m.buf = subclass(m.buf)
+            return m
+
+        def _test(module_subclass=None, sd_subclass=None):
+            m = _create_model(module_subclass)
+            sd = _create_model(sd_subclass).state_dict()
+            # deepcopy as swap_tensors might destructively modify state dict
```
Comment (albanD) [inline test/nn/test_load_state_dict.py:487]: Ho that is indeed an interesting side effect...

File: test/nn/test_load_state_dict.py [hunk 8]
```diff
@@ -362,7 +365,162 @@ def test_load_state_dict_warn_assign(self):
...
+            self.assertTrue(type(m.weight) is weight_type)
+            self.assertTrue(type(m.buf) is buf_type)
+
+        subclasses = [None, MyLoadTensor, MyWrapperLoadTensor]
+        for m_s, sd_s in product(subclasses, subclasses):
+            _test(m_s, sd_s)
+
+
+# run tests witsh swap enabled/disabled
+def instantiate_parametrize_with_swap(cls):
```
Comment (albanD) [inline test/nn/test_load_state_dict.py:512]: Why do you need this here? Can't it be in the utils or included in the parametrized test generation?

File: torch/_tensor.py [hunk 9]
```diff
@@ -710,6 +710,64 @@ def share_memory_(self):
...
+
+        ..note::
+            If both ``self`` and ``other`` have ``__torch_function__`` implemented, this
+            method will take precedence over :meth:`~Tensor.module_load_to`.
+
+        Args:
+            other (Tensor): value in state dict with key corresponding to ``self``
+
+        """
+        if has_torch_function_unary(self):
```
Comment (albanD) [inline torch/_tensor.py:738]: You should be checking both self and other here. This is quite meh

File: test/nn/test_load_state_dict.py (hunk 3 above)
Comment (mikaylagawarecki) [reply test/nn/test_load_state_dict.py:385]: `MyWrapperLoadTensor._data`, I named the attribute of the wrapper subclass as such, but can rename

File: test/nn/test_load_state_dict.py (hunk 2 above)
Comment (mikaylagawarecki) [reply test/nn/test_load_state_dict.py:370]: I think  `new_empty` needs to be implemented as `__deepcopy__` goes through this path for non-wrapper

https://github.com/pytorch/pytorch/blob/9fe3693bbb9494f57febebe0d6aefa9da081b16a/torch/_tensor.py#L172-L188

The override seemed to error if within the `__torch_function__` but that might just be a bug on my end, will retry.

File: test/nn/test_load_state_dict.py (hunk 6 above)
Comment (mikaylagawarecki) [reply test/nn/test_load_state_dict.py:406]: Hmmm one is a wrapper subclass and one is not though, can I still inherit from the non-wrapper MyLoadTensor

File: test/nn/test_load_state_dict.py (hunk 5 above)
Comment (mikaylagawarecki) [reply test/nn/test_load_state_dict.py:424]: I think this was wrong, `new_empty` should only be needed for non-wrapper subclasses, my bad

File: test/nn/test_load_state_dict.py (hunk 3 above)
Comment (albanD) [reply test/nn/test_load_state_dict.py:385]: Ho my bad. I though it was a field on the base Tensor.

File: test/nn/test_load_state_dict.py (hunk 8 above)
Comment (mikaylagawarecki) [reply test/nn/test_load_state_dict.py:512]: Do you mean move this function to a test utils file?

I want to run each test within the parametrized class with two different decorators that contain context managers that set `swap`.

I did not want to include it in the parametrized test generation (`instantiate_parametrized_tests`) as that seemed too "large blast" and I don't think we need to parametrize that many tests with swap True and False

I think using `@parametrize` here would require explicitly using the context manager in every test within the test class which I preferred not to do :P

I agree that my solution here is not clean though, do you know of any utility that would achieve this?

File: test/nn/test_load_state_dict.py (hunk 7 above)
Comment (mikaylagawarecki) [reply test/nn/test_load_state_dict.py]: The current state is that 
- swap_tensors does not destructively modify state dict
-  in the case where `param` or `sd[key]` has `__torch_function__` handler for `module_load` if `id(param.module_load(sd[key])) == id(sd[key])`, `param`'s `THPVariable->cdata` will point to the same `TensorImpl` as `sd[key]`'s `THPVariable->cdata`
- I added a warning in the `nn.Module.load_state_dict` docs rearding this 

After thinking about it more, I am not sure that we want to `clone()` the `sd[key]` to preserve behavior of the non-swap path. I think it is unlikely that people use the `state_dict` after `load_state_dict` but curious for your feedback on this!

File: test/nn/test_load_state_dict.py (hunk 8 above)
Comment (mikaylagawarecki) [reply test/nn/test_load_state_dict.py]: Oops ok changed this to a `@swap` `_TestParametrizer` and decorated each function

File: test/nn/test_load_state_dict.py [hunk 10]
```diff
@@ -8,17 +10,29 @@
...
     TEST_NUMPY, IS_WINDOWS, skipIfTorchDynamo, instantiate_parametrized_tests, \
-    run_tests
+    run_tests, wrapSwapTensorsTest, skipIfCrossRef, _TestParametrizer
+from torch.utils._pytree import tree_map
 
 if TEST_NUMPY:
     import numpy as np
 
 
+class swap(_TestParametrizer):
```
Comment (albanD) [inline test/nn/test_load_state_dict.py:20]: Why isn't this living in common_utils with wrapSwapTensorsTest?

File: test/nn/test_load_state_dict.py [hunk 11]
```diff
@@ -362,7 +386,123 @@ def test_load_state_dict_warn_assign(self):
...
+                    return cls(src._data)
+                return cls(src)
+        else:
+            return src
+
+    if func is torch.Tensor.module_load:
+        return module_load(*args, **kwargs)
+    else:
+        with torch._C.DisableTorchFunctionSubclass():
+            # detach must return instance of same subclass for nn.Parameter()
```
[the remark this answers is not in the document]
Comment (albanD) [reply test/nn/test_load_state_dict.py:412]: Is MyWrapperLoadTensor actually using this function though?

File: torch/nn/modules/module.py [hunk 12]
```diff
@@ -2101,6 +2114,13 @@ def load_state_dict(self, state_dict: Mapping[str, Any],
             If :attr:`assign` is ``True`` the optimizer must be created after
             the call to :attr:`load_state_dict`.
 
+        .. warning::
```
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply torch/nn/modules/module.py:2117]: This behavior differs from the default/might be surprising, so I thought it might be worth mentioning.

It is probably unlikely that the user will use the state_dict after loading though, so maybe it is okay to omit this

File: test/nn/test_load_state_dict.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply test/nn/test_load_state_dict.py:412]: My understanding is when creating `Parameter(MyWrapperLoadTensor)`

`MyWrapperLoadTensor.detach()` would call into `MyLoadTensor.__torch_function__` which would then go into the `__torch_dispatch__` for `MyWrapperLoadTensor.detach()`, so I do think it will go into this, yes.

File: torch/nn/modules/module.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (albanD) [reply torch/nn/modules/module.py:2117]: The default is to .copy_() into what is here right now. So yes the new behavior differs from it but that is much broader than what you mention here.
Also I think the flag setter is the right place to explain what the flag does.

File: test/nn/test_load_state_dict.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (albanD) [reply test/nn/test_load_state_dict.py:402]: Ok!

File: torch/__future__.py [hunk 13]
```diff
@@ -22,28 +25,41 @@ def set_overwrite_module_params_on_conversion(value: bool) -> None:
...
 
     #. ``module.{device}()`` (e.g. ``module.cuda()``) for moving a module between devices
     #. ``module.{dtype}()`` (e.g. ``module.float()``) for converting a module to a different dtype
        (for converting a module to a different dtype)
     #. ``module.to()``
+    #. ``module.load_state_dict(state_dict)``
+
+    The semantics for :meth:`~nn.Module.load_state_dict` when this is set are as follows:
+
+    #. ``sd['key']`` is transformed via :meth:`~torch.Tensor.module_load`
```
Comment (albanD) [inline torch/__future__.py:55]: Very clear explanation.
I don't think `sd['key']` is very meaningful though?

**Review by albanD — APPROVED** [review]
Small nit but SGTM!

File: torch/__future__.py (hunk 13 above)
Comment (mikaylagawarecki) [inline torch/__future__.py:55]:
```suggestion
    #. ``state_dict['key']`` is transformed via :meth:`~torch.Tensor.module_load`
```

File: torch/__future__.py [hunk 14]
```diff
@@ -22,28 +25,41 @@ def set_overwrite_module_params_on_conversion(value: bool) -> None:
...
+       (i.e. ``res = param.module_load(sd['key'])``)
```
Comment (mikaylagawarecki) [inline torch/__future__.py:56]:
```suggestion
    #. For each parameter/buffer, its corresponding``sd['key']`` is transformed via 
       :meth:`~torch.Tensor.module_load` (i.e. ``res = param.module_load(state_dict['key'])``)
```

File: torch/__future__.py [hunk 15]
```diff
@@ -22,28 +25,41 @@ def set_overwrite_module_params_on_conversion(value: bool) -> None:
...
 
     #. ``module.{device}()`` (e.g. ``module.cuda()``) for moving a module between devices
     #. ``module.{dtype}()`` (e.g. ``module.float()``) for converting a module to a different dtype
        (for converting a module to a different dtype)
     #. ``module.to()``
+    #. ``module.load_state_dict(state_dict)``
+
+    The semantics for :meth:`~nn.Module.load_state_dict` when this is set are as follows:
+
+    #. For each parameter/buffer, its corresponding``sd['key']`` is transformed via
```
Comment (mikaylagawarecki) [inline torch/__future__.py:55]:
```suggestion
    #. For each parameter/buffer, its corresponding``state_dict['key']`` is transformed via 
```
